### PR TITLE
Better performance for MulticastDelegate.Equals

### PR DESCRIPTION
**Pull request by jamesqo**
cc: @jkotas

### CODE REVIEW

**Comment by jkotas** [comment]
The actual delegate types inherit from MulticastDelegate, not Delegate. If you would like to improve the fast path for comparing identical delegate instances, it should be better done in MulticastDelegate.Equals. Results from a small microbenchmark would be nice.

**Comment by jamesqo** [comment]
@jkotas I see. I googled about `MutlicastDelegate` and came across this StackOverflow question: http://stackoverflow.com/a/4711164

If `System.Delegate` can't be inherited from in C# and the CLR is incapable of loading such a type, then why not just have the whole `Delegate.Equals` (and other methods, like `Delegate.GetInvocationList`) throw a `NotImplemented/SupportedException` or `Contract.Assert(false)`?

**Comment by jamesqo** [comment]
Here are the numbers for the new `MulticastDelegate.Equals`: [old](https://gist.github.com/jamesqo/0c3fdd2ad4482f48bfa0c3f0093ce104) / [new](https://gist.github.com/jamesqo/7becdac1ea2769a373639c7668cdca1d) / [source code](https://gist.github.com/jamesqo/7b4b795a62cffac011e995ebb3abbed0)

I added a `JitHelpers.UnsafeCast` after the `InternalEqualTypes` since we know the other object has to be a `MulticastDelegate`, but it doesn't seem to be having much of an impact in the case where the delegates are the same method, but not reference-equal. (I don't know the exact codepath it goes down in the method, so perhaps the latter parts have more of an overhead making the cast insignificant.)

For the delegates being reference-equal, there is a 10x speedup.

**Comment by stephentoub** [comment]
> For the delegates being reference-equal

What about when they're not?

**Comment by stephentoub** [comment]
> there is a 10x speedup


I'm confused. Comparing the old and new results you posted, reference equality looks appx the same between them, and method equality appears to regress 10%. What am I misunderstanding?

**Comment by jamesqo** [comment]
@stephentoub Whoops! Looks like I was actually testing the new code twice, I looked at the wrong copy of `System.Private.CoreLib.ni.dll` in ILSpy. Will post results for the old implementation soon.

**Comment by jamesqo** [comment]
I've updated the old results. It looks like the new changes actually do make it 10x faster, and have a 10% improvement for the method equality case.

File: src/mscorlib/src/System/MulticastDelegate.cs [hunk 1]
```diff
@@ -103,11 +104,22 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
...
-                return false;      
-            MulticastDelegate d = obj as MulticastDelegate;
-            if (d == null)
+            if (obj == null)
                 return false;
+            if (object.ReferenceEquals(this, obj))
+                return true;
+            
+            // TODO: See if this can be changed to GetType() == obj.GetType()
+            // without negatively impacting performance. GetType is a JIT
```
Comment (jkotas) [inline src/mscorlib/src/System/MulticastDelegate.cs]: The GetType JIT intrinsic is only optimized in specific usage patterns. This is not one of them.

Comment (jamesqo) [inline src/mscorlib/src/System/MulticastDelegate.cs]: @jkotas I removed the TODO. (It seems kinda unintuitive that you have to define a new function/ P/Invoke to get faster performance than what's written 99% of the time, but whatever.)

**Comment by jkotas** [comment]
LGTM. Thanks!

**Comment by jkotas** [comment]
@jamesqo Could you please make a matching fix in CoreRT as well?

**Pull request merged**